drv/cosmo-seq-server/src/main.rs

```diff
@@ -588,6 +593,127 @@ impl ServerImpl {
+            sys_api::IrqControl::Enable,
+        );
+    }
+
+    fn disable_sequencer_interrupts(&mut self) {
+        self.seq.ier.modify(|m| {
+            m.set_fanfault(false);
+            m.set_thermtrip(false);
+            m.set_smerr_assert(false);
+            m.set_a0mapo(false);
+            m.set_nicmapo(false);
+            m.set_amd_pwrok_fedge(false);
+            m.set_amd_rstn_fedge(false);
+        });
+        let _ = self.sys.gpio_irq_control(
+            notifications::SEQ_IRQ_MASK,
+            sys_api::IrqControl::Disable,
+        );
+    }
+
+    fn handle_sequencer_interrupt(&mut self) {
+        let ifr = self.seq.ifr.view();
+
+        let state = self.log_state_registers();
+        ringbuf_entry!(Trace::SequencerInterrupt {
+            our_state: self.state,
+            seq_state: state.seq,
+            ifr,
+        });
+
+        enum InternalAction {
+            Reset,
+            ThermTrip,
+            Smerr,
+            Mapo,
+        }
+
+        // We check these in lowest to highest priority. The expectation is
+        // we probably(?) won't see multiple of these set at a time but
+        // it's important to account for that case;
+
+        let mut action = InternalAction::Reset;
+
+        if ifr.amd_pwrok_fedge || ifr.amd_rstn_fedge {
+            let rstn = self.seq.amd_reset_fedges.counts();
+            let pwrokn = self.seq.amd_pwrok_fedges.counts();
+            // Reset edge interrupts flags
+            self.seq.ifr.modify(|h| {
+                h.set_amd_pwrok_fedge(false);
+                h.set_amd_rstn_fedge(false);
+            });
+
+            ringbuf_entry!(Trace::ResetCounts { rstn, pwrokn });
+            action = InternalAction::Reset;
+            // counters are cleared in the A2 -> A0 transition
+            // host_sp_comms will be notified of this change and will
+            // call back into this task to reboot the system (going to
+            // A2 then back into A0)
+        }
+
+        if ifr.thermtrip {
+            self.seq.ifr.modify(|h| h.set_thermtrip(false));
+            ringbuf_entry!(Trace::Thermtrip);
+            action = InternalAction::ThermTrip;
+            // Reat place for an ereport?
+        }
+
+        if ifr.a0mapo {
+            self.log_pg_registers();
+            self.seq.ifr.modify(|h| h.set_a0mapo(false));
+            ringbuf_entry!(Trace::A0MapoInterrupt);
+            action = InternalAction::Mapo;
+            // Great place for an ereport?
+        }
+        if ifr.smerr_assert {
+            self.seq.ifr.modify(|h| h.set_smerr_assert(false));
+            ringbuf_entry!(Trace::SmerrInterrupt);
+            action = InternalAction::Smerr;
+            // Great place for an ereport?
+        }
+        // Fan Fault is unconnected
+        // NIC MAPO is unconnected
+
+        match action {
+            InternalAction::Reset => {
+                self.set_state_internal(PowerState::A0Reset);
```

Comment:
Note that this is a terminal state (for now)?

(also relevant for `Mapo` and `Smerr`)